core/transaction_pool/impl/transaction_pool_impl.cpp

```diff
@@ -0,0 +1,190 @@
+
+namespace kagome::transaction_pool {
+
+  TransactionPoolImpl::TransactionPoolImpl(
+      std::unique_ptr<PoolModerator> moderator, Limits limits,
+      common::Logger logger)
+      : logger_{std::move(logger)},
+        moderator_{std::move(moderator)},
+        limits_{limits} {}
+
+  outcome::result<void> TransactionPoolImpl::submitOne(Transaction t) {
+    return submit({t});
+  }
+
+  outcome::result<void> TransactionPoolImpl::submit(
+      std::vector<Transaction> ts) {
+    for (auto &t : ts) {
+      auto hash = t.hash.toHex();
+      if (imported_hashes_.find(hash) != imported_hashes_.end()) {
+        return TransactionPoolError::ALREADY_IMPORTED;
+      }
+      for (auto &tag : t.provides) {
+        provided_tags_.insert({tag, t.hash});
+      }
+      imported_hashes_.insert(hash);
+      waiting_queue_.push_back(WaitingTransaction{std::move(t)});
+    }
+
+    updateReady();
+
+    enforceLimits();
+
+    return outcome::success();
+  }
+
+  std::vector<Transaction> TransactionPoolImpl::getReadyTransactions() {
+    std::vector<Transaction> ready(ready_queue_.size());
+    std::transform(ready_queue_.begin(), ready_queue_.end(), ready.begin(),
+                   [](auto &rtx) { return rtx.second; });
+    return ready;
+  }
+
+  std::vector<Transaction> TransactionPoolImpl::removeStale(
+      const primitives::BlockId &at) {
+    BlockNumber number;
+    if (at.type() == typeid(BlockNumber)) {
+      number = boost::get<BlockNumber>(at);
+    } else {
+      number = 0;
+      // TODO(Harrm) Figure out how to obtain block number from block hash
+    }
+
+    std::vector<Transaction> removed;
+
+    auto w_border =
+        std::stable_partition(waiting_queue_.begin(), waiting_queue_.end(),
+                              [this, number](auto &&tx) {
+                                return moderator_->banIfStale(number, tx);
+                              });
+    std::move(waiting_queue_.begin(), w_border, std::back_inserter(removed));
+    waiting_queue_.erase(waiting_queue_.begin(), w_border);
+
+    for (auto &[hash, ready_tx] : ready_queue_) {
+      if (moderator_->banIfStale(number, ready_tx)) {
+        removed.push_back(ready_tx);
+        ready_queue_.erase(hash);
+      }
+    }
+
+    moderator_->updateBan();
+
+    return removed;
+  }
+
+  void TransactionPoolImpl::updateReady() {
+    auto is_ready = [this](auto &&tx) {
+      return std::all_of(
+          tx.requires.begin(), tx.requires.end(), [this](auto &&tag) {
+            return provided_tags_.find(tag) != provided_tags_.end();
+          });
+    };
+    for (auto it = waiting_queue_.begin(); it != waiting_queue_.end();) {
+      if (is_ready(*it)) {
+        Transaction tx = *it;
+        ReadyTransaction rtx{tx, {}};
+        updateUnlockingTransactions(rtx);
+        ready_queue_.insert({rtx.hash, rtx});
+        auto old_it = it;
+        it++;
+        waiting_queue_.erase(old_it);
+      } else {
+        it++;
+      }
```

Comment:
it seems you can do 

```
if(is_ready(*it)){
  ...
}

++it

```

Reply:
I think that if I erase before incrementing 'it', it may become invalidated, so I prefer to move it to the next element first to make sure it'll stay valid.

Reply:
aw, I see. 
as I remember, erase increments iterator, so it should be 
```
it .= waiiting_quque_.erase(it);
```
(i may be wrong)